infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/batchprocessor/protocolbatchprocessor/ProtocolBatchProcessor.java

```diff
@@ -0,0 +1,220 @@
+import org.corfudb.runtime.view.AddressSpaceView;
+import org.corfudb.runtime.view.ReadOptions;
+import org.corfudb.util.retry.ExponentialBackoffRetry;
+import org.corfudb.util.retry.IRetry;
+import org.corfudb.util.retry.RetryNeededException;
+
+import java.time.Duration;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import static lombok.Builder.Default;
+import static org.corfudb.infrastructure.log.statetransfer.batch.TransferBatchResponse.TransferStatus.FAILED;
+
+/**
+ * A transferBatchRequest processor that transfers non-committed addresses one transferBatchRequest at a time
+ * via a replication protocol. A non-committed address is the one that belongs to the
+ * part of an address space that is inconsistent.
+ */
+@Slf4j
+@Builder
+public class ProtocolBatchProcessor implements StateTransferBatchProcessor {
+
+    /**
+     * Configurations for the retry logic.
+     */
+    @Default
+    private final int maxReadRetries = 3;
+    @Default
+    private final Duration maxRetryTimeout = Duration.ofSeconds(10);
+    @Default
+    private final float randomFactorBackoff = 0.5f;
+    @Default
+    private final AtomicInteger maxWriteRetries = new AtomicInteger(3);
+    @Default
+    private final Duration writeSleepDuration = Duration.ofMillis(500);
+    /**
+     * Default read options for the replication protocol read.
+     */
+    @Getter
+    private final ReadOptions readOptions = ReadOptions.builder()
+            .waitForHole(true)
+            .clientCacheable(false)
+            .serverCacheable(false)
+            .build();
+
+    @Getter
+    private final LogUnitClient logUnitClient;
+
+    @Getter
+    private final AddressSpaceView addressSpaceView;
+
+    @Override
+    public CompletableFuture<TransferBatchResponse> transfer(TransferBatchRequest transferBatchRequest) {
+        return readRecords(transferBatchRequest, 0)
+                .thenApply(records ->
+                        writeRecords(records, logUnitClient, maxWriteRetries, writeSleepDuration))
+                .exceptionally(error -> TransferBatchResponse
+                        .builder()
+                        .status(FAILED)
+                        .causeOfFailure(Optional.of(new StateTransferBatchProcessorException(error)))
+                        .build()
+                );
+    }
+
+    /**
+     * Reads data entries by utilizing the replication protocol. If there are errors, retries.
+     *
+     * @param transferBatchRequest The transferBatchRequest of addresses to be read.
+     * @param retries              The number of retries.
+     * @return A result of reading records.
+     */
+    CompletableFuture<ReadBatch> readRecords(TransferBatchRequest transferBatchRequest, int retries) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use explicit scoping instead of the default package private level](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651453)